Reland "Construct NJT without graph breaks"

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132814
* #132414
* __->__ #133196
* #133145

This reverts commit 154d40ca488e6979ce9c2de89d8a35b53129ebea.

and adds changes from https://github.com/pytorch/pytorch/pull/133061

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

soulitzer committed:
Revert "Revert "Construct NJT without graph breaks""

This reverts commit 154d40ca488e6979ce9c2de89d8a35b53129ebea.

[ghstack-poisoned]

soulitzer committed:
Update on "Reland "Construct NJT without graph breaks""


This reverts commit 154d40ca488e6979ce9c2de89d8a35b53129ebea.

and adds changes from https://github.com/pytorch/pytorch/pull/133061

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -552,27 +565,31 @@ def _memo_epoch(self, obj: FakeTensor) -> str:
             setattr(obj, self._memo(obj), value)
-            setattr(obj, self._memo_vc(obj), obj._version)
+            if not self._no_vc:
+                setattr(obj, self._memo_vc(obj), obj._version)
```
Changes to SymIntMemoDescriptor

soulitzer commented on test/dynamo/test_subclasses.py:
```diff
@@ -1754,6 +1755,417 @@ def binary(nt1, nt2):
+        def fn(n):
+            return n * 2
+
+        torch.compile(fn, backend="eager")(nt)
```
New test

soulitzer committed:
Update on "Reland "Construct NJT without graph breaks""


This reverts commit 154d40ca488e6979ce9c2de89d8a35b53129ebea.

and adds changes from https://github.com/pytorch/pytorch/pull/133061

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -531,8 +531,21 @@ class FakeTensorConfig:
+    ) -> None:
+        assert is_unbacked is not None
+        self._is_unbacked = is_unbacked
+        self._no_vc = no_vc
```
no_vc is the operational description here, but it might be more clear if it was something like `preserve_across_mutation: bool = False`

ezyang replied:
Or even better, just have a `nested_int` flag that controls both the is unbacked and no vc simultaneously, since that's the only use case right now. Not a big deal though.

ezyang reviewed (approved):
only rereviewed new bits

soulitzer commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -531,8 +531,21 @@ class FakeTensorConfig:
+    _no_vc: bool
+
+    def __init__(
+        self, *, is_unbacked: Optional[bool] = None, no_vc: bool = False
```
I guess it means do not-`preserve_across_epochs` lol
But for simplicity, just going to have a single `nested_int` flag per the other comment.

soulitzer committed:
Update on "Reland "Construct NJT without graph breaks""


This reverts commit 154d40ca488e6979ce9c2de89d8a35b53129ebea.

and adds changes from https://github.com/pytorch/pytorch/pull/133061

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge
